File: src/theWheelView/Elliptangle.cpp
/// <remarks>
/// Elliptangle.cpp
/// Implementation of the Elliptangle RadialShape class.
///
/// Copyright (C) 2006-2007 Derek G Lane
/// </remarks>

#include "StdAfx.h"

// class declaration
#include "Elliptangle.h"

namespace theWheel2007 {

/////////////////////////////////////////////////////////////////////////////
void 
	Elliptangle::SetExtentEllipt(CExtent<2, REAL> rectInner, REAL ellipt)
		// sets the extents (both inner and outer)
{
	// set inner extent
	InnerRect = rectInner;

	// calculate outer extent
	rectInner.Deflate(
		-(1.0f - ellipt) * rectInner.GetSize(0) / 2.0f, 
		-(1.0f - ellipt) * rectInner.GetSize(1) / 2.0f,
		-(1.0f - ellipt) * rectInner.GetSize(0) / 2.0f, 
		-(1.0f - ellipt) * rectInner.GetSize(1) / 2.0f);
	OuterRect = rectInner;

	// now calculate the supporting extents

	// calculate the fat ellipse extent

	// compute the fat ellipse's b
	REAL ry = InnerRect.GetSize(1) / OuterRect.GetSize(1); 
	REAL by = sqrt(InnerRect.GetSize(0) * InnerRect.GetSize(0) * 0.25f
						/ (1.0f - ry * ry));

	// compute the fat ellipse's rectangle
	m_extFatEllipse = OuterRect;
	m_extFatEllipse.Deflate(m_extFatEllipse.GetSize(0) / 2.0f, 0, m_extFatEllipse.GetSize(0) / 2.0f, 0);
	m_extFatEllipse.Deflate(-by, 0, -by, 0);


	// calculate the tall ellipse extent

	// compute the tall ellipse's b
	REAL rx = InnerRect.GetSize(0) / OuterRect.GetSize(0); 
	REAL bx = sqrt(InnerRect.GetSize(1) * InnerRect.GetSize(1) * 0.25f
						/ (1.0f - rx * rx));

	// compute the tall ellipse's rectangle
	m_extTallEllipse = OuterRect;
	m_extTallEllipse.Deflate(0, m_extTallEllipse.GetSize(1) / 2.0f, 0, m_extTallEllipse.GetSize(1) / 2.0f);
	m_extTallEllipse.Deflate(0, -bx, 0, -bx);


	// now set up discont list

	// compute the angle dividing the sections of the elliptangle
	REAL sectionAngle = atan2(OuterRect.GetSize(1), OuterRect.GetSize(0));

	// determine whether angle is in fat or tall ellipse
	m_arrDiscont.resize(4); // = gcnew array<REAL>(4);
	m_arrDiscont[0] = sectionAngle;
	m_arrDiscont[1] = PIf - sectionAngle;
	m_arrDiscont[2] = PIf + sectionAngle;
	m_arrDiscont[3] = 2.0f * PIf - sectionAngle;

}	// Elliptangle::SetExtentEllipt


//////////////////////////////////////////////////////////////////////////
CVectorD<2, REAL>
	Elliptangle::Eval(REAL theta)
		// helper function to evaluate an elliptangle
{
	CVectorD<2, REAL> vPoint;

	// check theta range
	if (theta < 0.0f)
	{
		vPoint = Eval(theta + 2.0f * PIf);
	}
	else if (2.0f * PIf < theta)
	{
		vPoint = Eval(theta - 2.0f * PIf);
	}
	else
	{
		// compare tan(theta) to slope, to see which section we are in 
		REAL slope = OuterRect.GetSize(1) / OuterRect.GetSize(0);
		if (fabs(tan(theta)) >= fabs(slope))
		{
			// tangent greater than slope, so we are in fat ellipse domain
			vPoint[0] = m_extFatEllipse.GetSize(0) / 2.0f;
			vPoint[1] = m_extFatEllipse.GetSize(1) / 2.0f;
		}
		else
		{
			// tangent less than slope, so we are in tall ellipse domain
			vPoint[0] = m_extTallEllipse.GetSize(0) / 2.0f;
			vPoint[1] = m_extTallEllipse.GetSize(1) / 2.0f;
		}

		// compute theta_prime for the ellipse
		REAL sign = cos(theta) > 0 ? 1.0f : -1.0f;
		REAL theta_prime = atan2(sign * vPoint[0] * tan(theta), sign * vPoint[1]);

		vPoint[0] *= cos(theta_prime);
		vPoint[1] *= sin(theta_prime);
	}

	return vPoint;

}	// Elliptangle::Eval

//////////////////////////////////////////////////////////////////////////
void 
	Elliptangle::InvEvalY(REAL Y, REAL& angle1, REAL& angle2)
{
	// compute point on tall ellipse
	REAL a = m_extTallEllipse.GetSize(0) / 2.0f;
	REAL b = m_extTallEllipse.GetSize(1) / 2.0f;
	REAL X = a * sqrt(1.0f - Y * Y / (b * b));

	// see if the point is greater than tall slope
	REAL slope = OuterRect.GetSize(1) / OuterRect.GetSize(0);
	if (fabs(Y / X) >= fabs(slope))
	{
		// yes, so compute for fat ellipse
		a = m_extFatEllipse.GetSize(0)  / 2.0f;
		b = m_extFatEllipse.GetSize(1) / 2.0f;
		X = a * sqrt(1.0f - Y * Y / (b * b));
	}

	// assign the values
	angle1 = atan2(Y, X);
	angle2 = atan2(Y, -X);
}

}	// namespace theWheel2007

File: weel-app/src/geometry/Elliptangle.ts
import { RadialShape } from './RadialShape';
import { Extent2D } from './Extent2D';
import { Vector2D } from './Vector2D';

export class Elliptangle extends RadialShape {
    fatEllipseExtent: Extent2D;
    tallEllipseExtent: Extent2D;

    constructor(rectInner: Extent2D, ellipt: number) {
        super();
        this.fatEllipseExtent = new Extent2D();
        this.tallEllipseExtent = new Extent2D();
        this.setExtentEllipt(rectInner, ellipt);
    }

    setExtentEllipt(rectInner: Extent2D, ellipt: number): void {
        // Set inner extent
        this.innerRect = rectInner.clone();

        // Calculate outer extent
        const outerRect = rectInner.clone();
        outerRect.deflate(
            -(1.0 - ellipt) * rectInner.getSize(0) / 2.0,
            -(1.0 - ellipt) * rectInner.getSize(1) / 2.0,
            -(1.0 - ellipt) * rectInner.getSize(0) / 2.0,
            -(1.0 - ellipt) * rectInner.getSize(1) / 2.0
        );
        this.outerRect = outerRect;

        // Calculate the fat ellipse extent
        const ry = this.innerRect.getSize(1) / this.outerRect.getSize(1);
        const by = Math.sqrt(this.innerRect.getSize(0) * this.innerRect.getSize(0) * 0.25 / (1.0 - ry * ry));

        this.fatEllipseExtent = this.outerRect.clone();
        this.fatEllipseExtent.deflate(
            this.fatEllipseExtent.getSize(0) / 2.0, 0,
            this.fatEllipseExtent.getSize(0) / 2.0, 0
        );
        this.fatEllipseExtent.deflate(-by, 0, -by, 0);

        // Calculate the tall ellipse extent
        const rx = this.innerRect.getSize(0) / this.outerRect.getSize(0);
        const bx = Math.sqrt(this.innerRect.getSize(1) * this.innerRect.getSize(1) * 0.25 / (1.0 - rx * rx));

        this.tallEllipseExtent = this.outerRect.clone();
        this.tallEllipseExtent.deflate(
            0, this.tallEllipseExtent.getSize(1) / 2.0,
            0, this.tallEllipseExtent.getSize(1) / 2.0
        );
        this.tallEllipseExtent.deflate(0, -bx, 0, -bx);

        // Set up discontinuity list
        const sectionAngle = Math.atan2(this.outerRect.getSize(1), this.outerRect.getSize(0));
        this.discontinuities = [
            sectionAngle,
            Math.PI - sectionAngle,
            Math.PI + sectionAngle,
            2.0 * Math.PI - sectionAngle
        ];
    }

    eval(theta: number): Vector2D {
        let point = new Vector2D();

        // Check theta range
        if (theta < 0.0) {
            return this.eval(theta + 2.0 * Math.PI);
        } else if (2.0 * Math.PI < theta) {
            return this.eval(theta - 2.0 * Math.PI);
        } else {
            // Compare tan(theta) to slope to see which section we're in
            const slope = this.outerRect.getSize(1) / this.outerRect.getSize(0);

            let a, b;
            if (Math.abs(Math.tan(theta)) >= Math.abs(slope)) {
                // Tangent greater than slope, so we're in fat ellipse domain
                a = this.fatEllipseExtent.getSize(0) / 2.0;
                b = this.fatEllipseExtent.getSize(1) / 2.0;
            } else {
                // Tangent less than slope, so we're in tall ellipse domain
                a = this.tallEllipseExtent.getSize(0) / 2.0;
                b = this.tallEllipseExtent.getSize(1) / 2.0;
            }

            // Compute theta_prime for the ellipse
            const sign = Math.cos(theta) > 0 ? 1.0 : -1.0;
            const theta_prime = Math.atan2(sign * a * Math.tan(theta), sign * b);

            point.x = a * Math.cos(theta_prime);
            point.y = b * Math.sin(theta_prime);
        }

        return point;
    }

    invEvalY(y: number): { angle1: number, angle2: number } {
        // Compute point on tall ellipse
        let a = this.tallEllipseExtent.getSize(0) / 2.0;
        let b = this.tallEllipseExtent.getSize(1) / 2.0;
        let x = a * Math.sqrt(1.0 - y * y / (b * b));

        // See if the point is greater than tall slope
        const slope = this.outerRect.getSize(1) / this.outerRect.getSize(0);
        if (Math.abs(y / x) >= Math.abs(slope)) {
            // Yes, so compute for fat ellipse
            a = this.fatEllipseExtent.getSize(0) / 2.0;
            b = this.fatEllipseExtent.getSize(1) / 2.0;
            x = a * Math.sqrt(1.0 - y * y / (b * b));
        }

        // Return the angles
        return {
            angle1: Math.atan2(y, x),
            angle2: Math.atan2(y, -x)
        };
    }
}